### ISSUE TITLE
Compactor2 "Cold Compactions"

### ISSUE THREAD

**Issue by alamb** [influxdata/influxdb_iox#7125]
__Use case:__
In IOx, for maximum query performance the parquet files must be "non-overlapping" in time, otherwise any overlapping files must be deduplicated.

The compactor must tradeoff the effort required to create non overlapping files with the probability that new data will arrive that will overlap with some of the beautifully non-overlapping files. 

However, typically partitions go "cold" meaning that the probability of receiving a new write is low (according to some heuristic measurement of that probability). 

__Proposal:__
Design and Implement some sort of "cold" compaction:
1. Heuristic to decide when data is cold
2. Algorithm to arrange files in a non overlapping way

__Current behaviour:__
The compactor2 algorithm will attempt to balance query performance and potential future data loading, leaving partitions with medium sized, overlapping files (see https://github.com/influxdata/influxdb_iox/pull/7079#discussion_r1124915359

__Desired behaviour:__
Once a partition is "cold" then ideally the compactor should take a final pass through the partition and ensure that it has a minimum number of large, non overlapping files (subject to the various maximum file size constraints)

Without this feature, partitions from days or weeks ago with large amounts of data may still have non trivial amounts of overlapping data resulting in slow query performance. The "old" compactor had cold compactions, triggered by the time since last write. `compactor2` does not yet have this feature. 

__Alternatives considered:__
Describe other solutions or features you considered.

**Comment by alamb** [comment]
The compactor2 already has some notion of 'cold' based on this setting: aka it will not try and do 'hot' compactions on such partitions

```rust
    /// The compactor will only consider compacting partitions that
    /// have new parquet files created within this many minutes.
    #[clap(
        long = "compaction_partition_minute_threshold",
        env = "INFLUXDB_IOX_COMPACTION_PARTITION_MINUTE_THRESHOLD",
        default_value = "10",
        action
    )]
    pub compaction_partition_minute_threshold: u64,
``

**Comment by NGA-TRAN** [comment]
They sound good to me Carol

**Comment by Joe-Blount** [comment]
> new_file_at of no more than COLD_WINDOW_MINUTES = 60 * 3 = 180 minutes before the COLD_MINUTE_THRESHOLD, so 4 hours ago from now. 

Am I correct in thinking this means if we fail to complete cold compaction within 4h, it will no longer meet the criteria for cold compaction selection, so it will still be ignored?  In which case, it would still be true that if a cold partition doesn't complete its compaction within a specific time, we'll never compact it (its just that more time would be allowed before we permanently give up on it).

> My proposal would be to start with what compactor1 did, which was run hot compaction 4 times then run cold compaction 1 time.

From a balance of resources perspective, 4:1 sounds like a fine starting point.
But can't each of those "runs" take a long time?  (10s of minutes, an hour?)
That sounds like we stop processing hot partitions for a while and start processing cold ones.  From a query perspective on hot partitions, that would have similar effects as the compactor being down (like: OOMing, crash loop backoff, etc).  For the duration of the cold compactions, all the hot partitions will be accumulating L0s.  This might trigger the new alert Nga is adding for compaction falling behind.

**Comment by NGA-TRAN** [comment]
> Am I correct in thinking this means if we fail to complete cold compaction within 4h, it will no longer meet the criteria for cold compaction selection, so it will still be ignored? In which case, it would still be true that if a cold partition doesn't complete its compaction within a specific time, we'll never compact it (its just that more time would be allowed before we permanently give up on it).

Correct. The candidate selection can be thought more and it can be improved without affecting the rest of the algorithm.


> From a balance of resources perspective, 4:1 sounds like a fine starting point.
But can't each of those "runs" take a long time? (10s of minutes, an hour?)
That sounds like we stop processing hot partitions for a while and start processing cold ones. From a query perspective on hot partitions, that would have similar effects as the compactor being down (like: OOMing, crash loop backoff, etc). For the duration of the cold compactions, all the hot partitions will be accumulating L0s. This might trigger the new alert Nga is adding for compaction falling behind.

Correct. That is why I am no sure what is the best option for this.

Also, @carols10cents I do not know if you can handle this the way that if only one or a few threats are compacting the last  cold (or hot) partition candidates, can we still be able to fill the pool of new hot (or cold) partition candidates and continue with available threats. If we can do this and if we are currently (need to verify) never pick a partition in the new pool that is running, we will ensure to use the max resource

**Comment by Joe-Blount** [comment]
> What if, instead of doing 4 rounds of hot compaction and 1 round of cold, we have one (or a configurable number) thread that's always working on cold compaction?

I think that's better.  It might be well behaved, but I can image it causing issues with OOMing.

Cold partitions keep coming in when discussing OOMs and crash loop backoffs.  I'm concerned the list of cold partitions might be dominated by the high memory users.  If that's true... running one cold compaction at all times might mean it is a high memory user most of the time.  
In the mirror cluster when it has problems OOMing, we think its because random chance runs multiple high memory compactions at the same time.  It eventually gets lucky not selecting two high memory users at the same time, allowing it to get past the OOM loop.  So if the cold compaction is biased towards high memory users, that might "stack the deck" against the luck that gets the mirror cluster past its OOMs.  The end result would be increased chance of OOM crash loop backoffs.

But having said that, I don't have a better idea.  So maybe make sure the logs distinguish if its a hot or cold partition being compacted.  Hopefully we'll get to start making progress on the OOM strategy soon.

**Comment by NGA-TRAN** [comment]
I think if things works as expected, cold or hot compaction will consume similar memory. In the case where a few partition area so behind and lead to large cold compactions, they will dominate whether we do them in their own cold compaction cycle  or with the hot ones.
I think if cold compaction  take too much memory, we may want to focus on improving our algorithm to not let cold compaction take too much memory

**Comment by NGA-TRAN** [comment]
> I propose an alternate: Run a second compactor2 service in production that is responsible ONLY for cold compactions

If we want to do this, we need a mechanism the same partition is not compacted by the other service. It is possible with a careful design

**Comment by alamb** [comment]
> If we want to do this, we need a mechanism the same partition is not compacted by the other service. It is possible with a careful design

Perhaps the first version can use "optimistic concurrency control" -- namely take a snapshot of all the `ParquetFile`es  in the partition when the compactor determines what step to do next, and then as part of `commit` ensure that all those parquet files are unmodified. If any has been modified since the initial planning, discard all the work and try again

**Comment by NGA-TRAN** [comment]
> Perhaps the first version can use "optimistic concurrency control" -- namely take a snapshot of all the ParquetFilees in the partition when the compactor determines what step to do next, and then as part of commit ensure that all those parquet files are unmodified. If any has been modified since the initial planning, discard all the work and try again

I like this `optimistic concurrency control` approach. 

- The **snapshot** the compactor needs is a list of non-soft-delete parquet files of the partition. I think we only need the file ID and compaction level of the file.
- Right before catalog update, we verify if they are the same. If not, stop compacting that partition. If yes, update the catalog and re-capture the snapshot. This means before update, we need to read the catalog to verify nothing compacted but new L0 files ingested from Ingester is ok.
- Since a partition is looped to compact in several rounds, the update will happen at each round and the snapshot needs to re-capture right after update.

This needs itself is  complicated, isolated and big enough and I think if we decide to go with this, it should be done first in in its own PR.

**Comment by carols10cents** [comment]
And now I've added a "Current Plan" section to this issue's description and created some more sub-tasks that I think can be merged in separately.
